src/os/bluestore/bluestore_types.cc

```diff
@@ -335,6 +335,241 @@ ostream& operator<<(ostream& out, const bluestore_extent_ref_map_t& m)
   return out;
 }
 
+// bluestore_blob_use_tracker_t
+
+void bluestore_blob_use_tracker_t::allocate()
+{
+  assert(num_au != 0);
+  bytes_per_au = new uint32_t[num_au];
+}
+
+void bluestore_blob_use_tracker_t::fall_back_to_partial(uint32_t _num_au, uint32_t _au_size)
+{
+  assert(num_au == 0 && au_size == 0);
+  assert(_num_au != 0 && _au_size != 0);
+  assert(_num_au * _au_size >= total_bytes);
+
+  uint32_t total = total_bytes;
+  num_au = _num_au;
+  au_size = _au_size;
+  allocate();
+  for (size_t i = 0; i < num_au; ++i) {
+    bytes_per_au[i] = min(total, au_size);
+    total -= bytes_per_au[i];
+  }
+  assert(total == 0);
+}
+
+void bluestore_blob_use_tracker_t::get(
+  uint32_t offset, uint32_t length,
+  std::function<void(uint32_t*, uint32_t*)> fallback_cb)
+{
+  if (!num_au) {
+    if (offset == 0 && total_bytes == 0) {
+      total_bytes = length;
+      return;
+    }
+    uint32_t _au_size = 0, _num_au = 0;
+    fallback_cb(&_num_au, &_au_size);
+    if (_num_au && _au_size) { // we might be forbidden to fallback to partial
```

Comment:
I think we want to also check whether the _au_size is > than the blob size, and only do the fallback then.  Because there is no point in doing per-au counting when the blob is a single au (or, in the weird case where the min_alloc_size has changed, smaller).

(Thanks for the variable rename--this makes sense now :)